### PR TITLE
[179864] fix credo error: large abc size

### PR DESCRIPTION
**Pull request by irisTa56**
ticket: https://acsmine.tok.access-company.com/redmine/issues/179864

This commit passes `mix credo --strict` for the below three files:

* `lib/mix/gear_static_analysis.ex`
* `lib/web/template/precompiler.ex`
* `core/async_job/queue.ex`

Most of the changes are deconstruction of `cond` and `case` in a function by defining same-name functions distinguished by pattern-matching (like overloading); These changes are for reducing ABC size of functions.

### CODE REVIEW

File: core/async_job/queue.ex [hunk 1]
```diff
@@ -54,20 +54,41 @@ defmodule AntikytheraCore.AsyncJob.Queue do
...
+        {:ok, remove_broker(q1, pid)}
+        |> maintain_invariants_and_return(now_millis)
+      {{:cancel, job_id}                      , now_millis} ->
+        cancel_job(q1, job_id)
+        |> maintain_invariants_and_return(now_millis)
+      {:get_metrics                           , now_millis} ->
+        {metrics(q1), q1}
+        |> maintain_invariants_and_return(now_millis)
+      _                                                     ->
+        {:ok, q1} # failsafe: not to crash on unexpected command
```
Comment (ttokit) [inline core/async_job/queue.ex]: この辺りのパターンマッチは変更前に比べて各パターン間の処理の差異の可読性(というか視認性?)が失われてしまっている気がするので、悩ましいですね。。

File: core/async_job/queue.ex [hunk 2]
```diff
@@ -85,30 +106,14 @@ defmodule AntikytheraCore.AsyncJob.Queue do
     %__MODULE__{q | abandoned_jobs: []}
     |> move_jobs_running_too_long(now_millis)
   end
-  defp move_jobs_running_too_long(%__MODULE__{jobs:           jobs,
-                                              index_waiting:  index_waiting,
+  defp move_jobs_running_too_long(%__MODULE__{jobs:           _,
```
Comment (ttokit) [inline core/async_job/queue.ex]: 使わない値についてはマッチさせなくて良いと思います。

File: core/async_job/queue.ex (hunk 1 above)
Comment (irisTa56) [reply core/async_job/queue.ex]: @ttokit 
ご指摘ありがとうございます。
下記のような感じにすれば視認性を維持できると考えていますが、いかがでしょうか？
最終的に `maintain_invariants_and_return` に収束しているのも分かりやすいと思います。

```elixir
case cmd do
{{:add, job_key, job}                   , now_millis} -> {insert(q1, job_key, job),                         now_millis}
{{:fetch, broker}                       , now_millis} -> {fetch(q1, broker, now_millis),                    now_millis}
{{:remove_locked, job_key}              , now_millis} -> {{:ok, remove_locked(q1, job_key, now_millis)},    now_millis}
{{:unlock_for_retry, job_key}           , now_millis} -> {{:ok, unlock_for_retry(q1, job_key, now_millis)}, now_millis}
{{:remove_broker_from_waiting_list, pid}, now_millis} -> {{:ok, remove_broker(q1, pid)},                    now_millis}
{{:cancel, job_id}                      , now_millis} -> {cancel_job(q1, job_id),                           now_millis}
{:get_metrics                           , now_millis} -> { {metrics(q1), q1},                               now_millis}
_                                                     -> {{:ok, q1}} # failsafe: not to crash on unexpected command
end
|> case do
  {first_arg, now_millis} -> maintain_invariants_and_return(first_arg, now_millis)
  {failsafe}              -> failsafe
end
```

Comment (ttokit) [reply core/async_job/queue.ex]: 個人的には書いていただいた案が良いと思います！

File: core/async_job/queue.ex (hunk 2 above)
Comment (irisTa56) [reply core/async_job/queue.ex]: 申し訳ありません。
研修でやったところなのに。。。
修正いたしました。https://github.com/access-company/antikythera/pull/51/commits/fc6e820448960bb0e97310da0d884e43eed1b166

File: core/async_job/queue.ex [hunk 3]
```diff
@@ -54,20 +54,30 @@ defmodule AntikytheraCore.AsyncJob.Queue do
   @impl true
   defun new() :: t do
     set = :gb_sets.empty()
-    %__MODULE__{jobs: %{}, index_waiting: set, index_runnable: set, index_running: set, brokers_waiting: [], brokers_to_notify: [], abandoned_jobs: []}
+    %__MODULE__{jobs:              %{},
```
Comment (aMasakiTakahashi) [inline core/async_job/queue.ex]: Please write in [the same way](https://github.com/access-company/antikythera/blob/master/lib/web/conn.ex#L199) other codes are written.
* trailing comma
* position of braces
```
%__MODULE__{
  jobs:              %{},
  index_waiting:     set,
  index_runnable:    set,
  index_running:     set,
  brokers_waiting:   [],
  brokers_to_notify: [],
  abandoned_jobs:    [],
}
```

File: core/async_job/queue.ex [hunk 4]
```diff
@@ -125,28 +116,70 @@ defmodule AntikytheraCore.AsyncJob.Queue do
     end
   end
 
+  defp move_job_running_too_long(q, target_job_info, _now_millis, _threshold_time) when is_nil(target_job_info), do: q
```
Comment (aMasakiTakahashi) [inline core/async_job/queue.ex]: Following is more simple and general.
```
defp move_job_running_too_long(q, nil, _now_millis, _threshold_time), do: q
```

File: lib/mix/gear_static_analysis.ex [hunk 5]
```diff
@@ -75,75 +91,100 @@ defmodule Mix.Tasks.Compile.GearStaticAnalysis do
...
+    # We don't have to check `defimpl` without `for:`, as the enclosing module's name is enforced to be properly prefixed by gear name.
+    check_defimpl(concated_protocol_atoms, concated_mod_atoms, meta, file)
   end
 
+  defp check_ast_node({:use, meta, [{:__aliases__, _, atoms} | kw]}, file, _tool?) do
+    with_concatenated_module_atom(atoms, fn mod ->
+      check_use_within_module(mod, kw, meta, file)
+    end)
+  end
+
```
Comment (aMasakiTakahashi) [inline lib/mix/gear_static_analysis.ex]: Please remove blank line.
https://github.com/rrrene/elixir-style-guide#group-function-definitions

File: core/async_job/queue.ex (hunk 3 above)
Comment (irisTa56) [reply core/async_job/queue.ex]: I fixed it in https://github.com/access-company/antikythera/commit/43cce81c61461e12e362aa199679425fc58dcaa3

File: core/async_job/queue.ex (hunk 4 above)
Comment (irisTa56) [reply core/async_job/queue.ex]: I fixed it in https://github.com/access-company/antikythera/commit/43cce81c61461e12e362aa199679425fc58dcaa3

File: lib/mix/gear_static_analysis.ex (hunk 5 above)
Comment (irisTa56) [reply lib/mix/gear_static_analysis.ex]: Thank you for the reference!
I fixed it in https://github.com/access-company/antikythera/pull/51/commits/321f10e26d7dc5be21f956fe70833dc4ce32588e

**Comment by irisTa56** [comment]
I found that Credo can be disabled in specified lines by writing `credo:disable-for-lines:<count>` as a comment. Please see [this blog](http://trivelop.de/2017/06/01/credo-config-comments/) for more details.

Now, `Antikythera.TemplatePrecompiler.extract_vars_in_ast_node` stays simple!

File: core/async_job/queue.ex [hunk 6]
```diff
@@ -125,28 +118,70 @@ defmodule AntikytheraCore.AsyncJob.Queue do
...
-          []       -> q2
-          [b | bs] -> %__MODULE__{q2 | brokers_waiting: bs, brokers_to_notify: [b | brokers_to_notify]}
-        end
-        |> move_now_runnable_jobs_impl(now_millis)
-    end
+    target_job_info = take_smallest_with_earlier_timestamp(index_waiting, now_millis)
+    move_now_runnable_job(q, target_job_info, now_millis)
+  end
+
+  defp move_now_runnable_job(q, nil = _target_job_info, _now_millis), do: q
```
Comment (ttokit) [inline core/async_job/queue.ex]: We can infer the role of this argument from the argument name of the same function.
```suggestion
  defp move_now_runnable_job(q, nil, _now_millis), do: q
```

File: core/async_job/queue.ex [hunk 7]
```diff
@@ -125,28 +118,70 @@ defmodule AntikytheraCore.AsyncJob.Queue do
...
+                                  {{_, job_id} = job_key, index_waiting2} = _target_job_info,
+                                  now_millis) do
+    jobs2 = Map.update!(jobs, job_id, fn {j, t, :waiting} -> {j, t, :runnable} end)
+    index_runnable2 = :gb_sets.add(job_key, index_runnable)
+    q2 = %__MODULE__{q | jobs: jobs2, index_waiting: index_waiting2, index_runnable: index_runnable2}
+    move_first_waiting_broker(q2, brokers_waiting)
+    |> move_now_runnable_jobs_impl(now_millis)
+  end
+
+  defp move_first_waiting_broker(q, [] = _brokers_waiting), do: q
```
Comment (ttokit) [inline core/async_job/queue.ex]: Same as https://github.com/access-company/antikythera/pull/51#discussion_r300550170.
```suggestion
  defp move_first_waiting_broker(q, []), do: q
```

File: core/async_job/queue.ex [hunk 8]
```diff
@@ -125,28 +118,70 @@ defmodule AntikytheraCore.AsyncJob.Queue do
...
+    target_job_info = take_smallest_with_earlier_timestamp(index_waiting, now_millis)
+    move_now_runnable_job(q, target_job_info, now_millis)
+  end
+
+  defp move_now_runnable_job(q, nil = _target_job_info, _now_millis), do: q
+  defp move_now_runnable_job(%__MODULE__{jobs:            jobs,
+                                         index_runnable:  index_runnable,
+                                         brokers_waiting: brokers_waiting} = q,
+                                  {{_, job_id} = job_key, index_waiting2} = _target_job_info,
+                                  now_millis) do
```
Comment (ttokit) [inline core/async_job/queue.ex]: The alignment is misaligned.

File: core/async_job/queue.ex [hunk 9]
```diff
@@ -125,28 +118,70 @@ defmodule AntikytheraCore.AsyncJob.Queue do
     end
   end
 
+  defp move_job_running_too_long(q, nil = _target_job_info, _now_millis, _threshold_time), do: q
```
Comment (ttokit) [inline core/async_job/queue.ex]: Same as https://github.com/access-company/antikythera/pull/51#discussion_r300550170.
```suggestion
  defp move_job_running_too_long(q, nil, _now_millis, _threshold_time), do: q
```

File: core/async_job/queue.ex (hunk 6 above)
Comment (irisTa56) [reply core/async_job/queue.ex]: I fixed it https://github.com/access-company/antikythera/pull/51/commits/43b95be6693fa8736bf6caf8bd391781a4ccc141

File: core/async_job/queue.ex (hunk 7 above)
Comment (irisTa56) [reply core/async_job/queue.ex]: I fixed it https://github.com/access-company/antikythera/pull/51/commits/43b95be6693fa8736bf6caf8bd391781a4ccc141

File: core/async_job/queue.ex (hunk 9 above)
Comment (irisTa56) [reply core/async_job/queue.ex]: I fixed it https://github.com/access-company/antikythera/pull/51/commits/43b95be6693fa8736bf6caf8bd391781a4ccc141

File: core/async_job/queue.ex (hunk 8 above)
Comment (irisTa56) [reply core/async_job/queue.ex]: I fixed it https://github.com/access-company/antikythera/pull/51/commits/1fe049c9103c44dd7af89498437fff5dad0c9bf4

File: core/async_job/queue.ex (hunk 1 above)
Comment (irisTa56) [reply core/async_job/queue.ex]: I implemented the above suggestion in https://github.com/access-company/antikythera/pull/51/commits/fc6e820448960bb0e97310da0d884e43eed1b166

File: core/async_job/queue.ex [hunk 10]
```diff
@@ -54,20 +54,32 @@ defmodule AntikytheraCore.AsyncJob.Queue do
...
-      {:get_metrics                           , now_millis} -> {metrics(q1), q1}                                |> maintain_invariants_and_return(now_millis)
-      _                                                     -> {:ok, q1} # failsafe: not to crash on unexpected command
+      {{:add, job_key, job}                   , now_millis} -> {insert(q1, job_key, job),                         now_millis}
+      {{:fetch, broker}                       , now_millis} -> {fetch(q1, broker, now_millis),                    now_millis}
+      {{:remove_locked, job_key}              , now_millis} -> {{:ok, remove_locked(q1, job_key, now_millis)},    now_millis}
+      {{:unlock_for_retry, job_key}           , now_millis} -> {{:ok, unlock_for_retry(q1, job_key, now_millis)}, now_millis}
+      {{:remove_broker_from_waiting_list, pid}, now_millis} -> {{:ok, remove_broker(q1, pid)},                    now_millis}
+      {{:cancel, job_id}                      , now_millis} -> {cancel_job(q1, job_id),                           now_millis}
+      {:get_metrics                           , now_millis} -> { {metrics(q1), q1},                               now_millis}
```
Comment (ttokit) [inline core/async_job/queue.ex]:
```suggestion
      {:get_metrics                           , now_millis} -> {{metrics(q1), q1},                                now_millis}
```

Comment (irisTa56) [reply core/async_job/queue.ex]: I fixed it https://github.com/access-company/antikythera/pull/51/commits/3f4bc5fb27cc82dfbf23aca3959c93948eff71d5

File: core/async_job/queue.ex [hunk 11]
```diff
@@ -125,28 +118,70 @@ defmodule AntikytheraCore.AsyncJob.Queue do
     end
   end
 
+  defp move_job_running_too_long(q, nil, _now_millis, _threshold_time), do: q
+  defp move_job_running_too_long(%__MODULE__{jobs:          jobs,
+                                             index_waiting: index_waiting} = q,
+                                 {{_, job_id}, _} = target_job_info,
+                                 now_millis,
+                                 threshold_time) do
+    {j1, t, :running} = Map.fetch!(jobs, job_id)
```
Comment (ttokit) [inline core/async_job/queue.ex]: Now, `j1`'s numbering looks a bit strange, because `j2` moved to another function.
I just think the name `job` is good.

**Comment by ttokit** [comment]
LGTM.

File: core/async_job/queue.ex (hunk 11 above)
Comment (irisTa56) [reply core/async_job/queue.ex]: I fixed it in https://github.com/access-company/antikythera/pull/51/commits/ecfa650531ffbac8adfeabae5faaf2f43945ff1e

**Comment by irisTa56** [comment]
There seems to be something wrong in the scope of this PR.

```
Generated testgear app
** (ArgumentError) argument error
    :erlang.binary_to_existing_atom("Elixir.ExecutorPool.Id", :utf8)
    lib/mix/gear_static_analysis.ex:133: Mix.Tasks.Compile.GearStaticAnalysis.with_concatenated_module_atom/2
    lib/mix/gear_static_analysis.ex:89: anonymous fn/4 in Mix.Tasks.Compile.GearStaticAnalysis.check_module_body/4
    lib/util/macro_util.ex:11: anonymous fn/3 in Antikythera.MacroUtil.prewalk_accumulate/3
    (elixir) lib/macro.ex:282: anonymous fn/4 in Macro.do_traverse_args/4
    (elixir) lib/enum.ex:1418: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
    (elixir) lib/enum.ex:1418: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
Exited with code 1
```

I will revert the merge commit and investigate the problem.

**Pull request merged**